### PR TITLE
store cells in blocks for MixedTensor

### PR DESCRIPTION
**Pull request by arnej27959**
@bratseth or @lesters or @jonmv please review

note: two internal classes are no longer public;
I'm taking the bet that they aren't actually used outside the class itself.

### CODE REVIEW

File: vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java [hunk 1]
```diff
@@ -26,17 +26,44 @@ public class MixedTensor implements Tensor {
 
     /** The dimension specification for this tensor */
     private final TensorType type;
+    private final int blockSize; // aka dense subspace size
 
-    /** The list of cells in the tensor */
-    private final List<Cell> cells;
+    static class DenseBlock {
+        final TensorAddress sparseAddr;
+        final double[] cells;
```
Comment (jonmv) [inline vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java:35]: Do we want coarser numbers here too, later?

**Review by jonmv — DISMISSED** [review]
Nice!

**Review by baldersheim — DISMISSED** [review]
👍

File: vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java (hunk 1 above)
Comment (arnej27959) [reply vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java:35]: yes; I would like to have float and byte as options at least.

File: vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java [hunk 2]
```diff
@@ -26,17 +28,53 @@ public class MixedTensor implements Tensor {
...
+        if (this.blockSize < 1) {
+            throw new IllegalStateException("invalid dense subspace size " + blockSize);
+        }
+        long count = 0;
+        for (var block : this.cellBlocks) {
+            if (index.sparseMap.get(block.sparseAddr) != count) {
+                throw new IllegalStateException("map vs list mismatch");
+            }
+            if (block.cells.length != blockSize) {
+                throw new IllegalStateException("block length mismatch");
```
Comment (geirst) [inline vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java:71]: Consider outputting expected vs actual values so its easier to pinpoint the issue if this happens.

File: vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java [hunk 3]
```diff
@@ -26,17 +28,53 @@ public class MixedTensor implements Tensor {
...
+            }
+            ++count;
+        }
+        if (count != index.sparseMap.size()) {
+            throw new IllegalStateException("map vs list size mismatch");
```
Comment (geirst) [inline vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java:76]: Consider outputting expected vs actual values so its easier to pinpoint the issue if this happens.

File: vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java [hunk 4]
```diff
@@ -26,17 +28,53 @@ public class MixedTensor implements Tensor {
...
+        this.blockSize = index.denseSubspaceSize();
+        this.cellBlocks = List.copyOf(cellBlocks);
         this.index = index;
+        if (this.blockSize < 1) {
+            throw new IllegalStateException("invalid dense subspace size " + blockSize);
+        }
+        long count = 0;
+        for (var block : this.cellBlocks) {
+            if (index.sparseMap.get(block.sparseAddr) != count) {
+                throw new IllegalStateException("map vs list mismatch");
```
Comment (geirst) [inline vespajlib/src/main/java/com/yahoo/tensor/MixedTensor.java:68]: Consider outputting the sparse address so its easier to pinpoint the issue if this happens.

**Pull request merged**